test/core/prob_macro.jl

```diff
@@ -0,0 +1,78 @@
+    @testset "scalar" begin
+        @model demo(x) = begin
+            m ~ Normal()
+            x ~ Normal(m, 1)
+        end
+
+        mval = 3
+        xval = 2
+        iters = 1000
+
+        logprior = logpdf(Normal(), mval)
+        loglike = logpdf(Normal(mval, 1), xval)
+        logjoint = logprior + loglike
+
+        @test logprob"m = mval | model = demo" == logprior
+        @test logprob"m = mval | x = xval, model = demo" == logprior
+        @test logprob"x = xval | m = mval, model = demo" == loglike
+        @test logprob"x = xval, m = mval | model = demo" == logjoint
+
+        varinfo = Turing.VarInfo(demo(xval))
+        @test logprob"m = mval | model = demo, varinfo = varinfo" == logprior
+        @test logprob"m = mval | x = xval, model = demo, varinfo = varinfo" == logprior
+        @test logprob"x = xval | m = mval, model = demo, varinfo = varinfo" == loglike
+        varinfo = Turing.VarInfo(demo(missing))
+        @test logprob"x = xval, m = mval | model = demo, varinfo = varinfo" == logjoint
+
+        chain = sample(demo(xval), IS(), iters)
+        chain2 = Chains(chain.value, chain.logevidence, chain.name_map, NamedTuple())
+        lps = logpdf.(Normal.(vec(chain["m"].value), 1), xval)
+        @test logprob"x = xval | chain = chain" == lps
+        @test logprob"x = xval | chain = chain2, model = demo" == lps
+        varinfo = Turing.VarInfo(demo(xval))
+        @test logprob"x = xval | chain = chain, varinfo = varinfo" == lps
+        @test logprob"x = xval | chain = chain2, model = demo, varinfo = varinfo" == lps
+    end
+
+    @testset "vector" begin
+        n = 5
+        @model demo(x, n = n, ::Type{T} = Float64) where {T} = begin
+            m = Vector{T}(undef, n)
+            @. m ~ Normal()
+            @. x ~ Normal.(m, 1)
+        end
+        mval = rand(n)
+        xval = rand(n)
+        iters = 1000
+
+        logprior = sum(logpdf.(Normal(), mval))
+        like(m, x) = sum(logpdf.(Normal.(m, 1), x))
+        loglike = like(mval, xval)
+        logjoint = logprior + loglike
+
+        @test logprob"m = mval | model = demo" == logprior
+        @test logprob"x = xval | m = mval, model = demo" == loglike
+        @test logprob"x = xval, m = mval | model = demo" == logjoint
+
+        varinfo = Turing.VarInfo(demo(xval))
+        @test logprob"m = mval | model = demo, varinfo = varinfo" == logprior
+        @test logprob"x = xval | m = mval, model = demo, varinfo = varinfo" == loglike
+        # Currently, we cannot easily pre-allocate `VarInfo` for vector data
+
+        chain = sample(demo(xval), HMC(0.5, 1), iters)
+        chain2 = Chains(chain.value, chain.logevidence, chain.name_map, NamedTuple())
+        lps = like.([[chain["m[$i]"].value[j] for i in 1:n] for j in 1:iters], Ref(xval))
+        @test logprob"x = xval | chain = chain" == lps
+        @test logprob"x = xval | chain = chain2, model = demo" == lps
+        @test logprob"x = xval | chain = chain, varinfo = varinfo" == lps
```

Comment:
`p(x = xval | m = chain[i, :m], x = xval)` ? Does it make sense to combine the4 arguments for varinfo and chain? What would happen if varinfo contains `m = 0`? `p(x = xval | m = chain[i, :m], m = 0)`?

Reply:
Again `varinfo` is just for pre-allocation. This is just in case the user didn't save `varinfo` in `chain` for some reason (even though the default now is to save it) and wants to pass their own for some reason.

Reply:
Ok, but what happens in the case of `p(x = xval | model = demo, chain = chain, m = mval)`?

Reply:
The `mval` is ignored I believe.

Reply:
What is the preferred behavior here?

Reply:
I would say it’s not well defined what happens. But I think a sensible approach would be that `m = mval` overwrites what is in `chain`.

Reply:
I say we deal with these cases in a future PR. This use of the syntax is likely to be very rare anyways. I won't even document it in the getting started, just the advanced section of the docs. Sounds good?

Reply:
https://github.com/TuringLang/Turing.jl/issues/1039